src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -258,39 +277,48 @@ private void HandleChannelException(IConnection connection, Exception ex)
                 return;
             }
 
-            if (ShouldInvalidateServer(ex))
+            if (connection.Generation < _connectionPool.Generation)
+            {
+                return; // stale generation number
+            }
```

Comment:
From the former part of the pseudo code in the [spec](https://github.com/mongodb/specifications/commit/7dd9c008be6edaf65368787d5e91c2ac3aeb95b8#diff-62b4da2e38e7d2a9c1a2d92d315895bbR1401-R1411)
```
  def isStaleError(topologyDescription, response):
        currentServer = topologyDescription.servers[server.address]
        currentGeneration = currentServer.pool.generation
        generation = get connection generation from response
        if generation < currentGeneration:
            # Stale generation number.
            return True

        currentTopologyVersion = currentServer.topologyVersion
        # True if the current error's topologyVersion is greater than the server's
        return compareTopologyVersion(currentTopologyVersion, response.get("topologyVersion")) >= 0
```
Combined with this requirement:
https://github.com/mongodb/specifications/commit/7dd9c008be6edaf65368787d5e91c2ac3aeb95b8#diff-62b4da2e38e7d2a9c1a2d92d315895bbR1217-R1223
```
When processing a network or command error, clients MUST first check the
error's `generation number`_. If the error's generation number is equal to
the pool's generation number then error handling MUST continue according to
`Network error when reading or writing`_ or
`"not master" and "node is recovering"`_. Otherwise, the error is considered
stale and the client MUST NOT update any topology state.
(See `Why ignore errors based on CMAP's generation number?`_)
```

I'm open to the idea of changing this to an equality check to more closely follow the spec and moving the inequality check to `IsStaleStateChangeError`

Reply:
After some additional thought, I've applied the above changes so that we match the spec a bit more closely. I'm resolving this comment and moving the relevant bits to a non outdated line so they show up in a diff.